src/Storages/MergeTree/MergeTreeIndexUtils.cpp

```diff
@@ -44,4 +47,104 @@ ASTPtr buildFilterNode(const ASTPtr & select_query, ASTs additional_filters)
     return filter_node;
 }
 
+namespace
+{
+
+bool isTrivialCast(const ActionsDAG::Node & node)
+{
+    if (node.function_base->getName() != "CAST" || node.children.size() != 2 || node.children[1]->type != ActionsDAG::ActionType::COLUMN)
+        return false;
+
+    const auto * column_const = typeid_cast<const ColumnConst *>(node.children[1]->column.get());
+    if (!column_const)
+        return false;
+
+    Field field = column_const->getField();
+    if (field.getType() != Field::Types::String)
+        return false;
+
+    auto type_name = field.safeGet<String>();
+    return node.children[0]->result_type->getName() == type_name;
+}
+
+const ActionsDAG::Node & cloneFilterDAGNodeForIndexesAnalysis(
+    const ActionsDAG::Node & node,
+    ActionsDAG & new_dag,
+    std::unordered_map<const ActionsDAG::Node *, const ActionsDAG::Node *> cloned_nodes)
+{
+    auto it = cloned_nodes.find(&node);
+    if (it != cloned_nodes.end())
+        return *it->second;
+
+    const ActionsDAG::Node * res;
+    switch (node.type)
+    {
+        case (ActionsDAG::ActionType::INPUT):
+        {
+            res = &new_dag.addInput({node.column, node.result_type, node.result_name});
+            break;
+        }
+        case (ActionsDAG::ActionType::COLUMN):
+        {
+            String name;
+            if (const auto * column_const = typeid_cast<const ColumnConst *>(node.column.get());
+                column_const && column_const->getDataType() != TypeIndex::Function)
+            {
+                /// Re-generate column name for constant.
+                name = ASTLiteral(column_const->getField()).getColumnName();
```

Comment:
Is it hard to compare actual constants in the index analysis instead of generating a new name for it? It may be slow for big constants.

Reply:
Apparently in PK analysis (and in secondary index analysis as well, I guess) we compare expressions by name. So that there are lookups like `plus(x, 42)` where the full expression name is important.

Ideally, we should be able to do deep comparison (as we do for projection analysis), or at least calculate a hash from the subtree. But  it would require to re-implement a lot of code.